src/StructuredLogger/ObjectModel/IProjectOrEvaluation.cs

```diff
@@ -1,7 +1,77 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Configuration;
+
 namespace Microsoft.Build.Logging.StructuredLogger
 {
     public interface IProjectOrEvaluation
     {
-        string TargetFramework { get; set; }
+        public string AdornmentString { get; }
+
+        public string TargetFramework { get; set; }
+
+        public string Platform { get; set; }
+
+        public string Configuration { get; set; }
     }
-}
\ No newline at end of file
+
+    public static class IProjectOrEvaluationHelper
+    {
+        private class IProjectOrEvaluationComparer : IEqualityComparer<IProjectOrEvaluation>
+        {
+            public bool Equals(IProjectOrEvaluation x, IProjectOrEvaluation y)
+            {
+                return x.TargetFramework == y.TargetFramework
+                    && x.Platform == y.Platform
+                    && x.Configuration == y.Configuration;
+            }
+
+            public int GetHashCode(IProjectOrEvaluation obj)
+            {
+                int hashcode = 123456789;
+                hashcode = hashcode * 987654321 ^ (obj.TargetFramework == null ? 1 : obj.TargetFramework.GetHashCode());
+                hashcode = hashcode * 987654321 ^ (obj.Platform == null ? 1 : obj.Platform.GetHashCode());
+                hashcode = hashcode * 987654321 ^ (obj.Configuration == null ? 1 : obj.Configuration.GetHashCode());
+
+                return hashcode;
+            }
+        }
+
+        private const string seperator = ",";
+
+        private static IProjectOrEvaluationComparer comparer = new();
+
+        private static Dictionary<IProjectOrEvaluation, string> AdornmentStringCache = new(comparer);
+
+        public static string GetAdornmentString(this IProjectOrEvaluation proj)

```

Comment:
Not sure if the cache is the best solution here. How about you add AdornmentString on all implementations of IProjectOrEvaluation, and just have a property there that is calculated from the other three properties on demand? We could cache it there or calculate on demand every time.

Reply:
I am not 100% sure what you had in mind.  The cache does reduce the total number of strings as every project will have a new string for every view.

Reply:
Ah, sorry, I misunderstood how your cache works. I like it. All good.